### PR TITLE
Dennis2

### PR DESCRIPTION
**Pull request by rileyjsumner**

### CODE REVIEW

File: src/buffer.cpp [hunk 1]
```diff
@@ -44,7 +44,7 @@ BufMgr::BufMgr(std::uint32_t bufs)
  * arithmetic to ensure the clock remains in the acceptable range
  */
 void BufMgr::advanceClock() {
-    this->clockHand = (this->clockHand + 1) % this->numBufs
+    this->clockHand = (this->clockHand + 1) % this->numBufs;
```
Comment (rileyjsumner) [inline src/buffer.cpp:47]: ah shit I got semicoloned lol

File: src/buffer.cpp [hunk 2]
```diff
@@ -55,38 +55,42 @@ void BufMgr::allocBuf(FrameId& frame) {
...
-    // if buffer frame has valid page in it, remove entry from hash table
+    // check if the frame chosen for allocation is dirty
+    if (bufDescTable[clockHand].dirty) {
+        bufDescTable[clockHand].file.writePage(bufPool[clockHand]); // here
+    }
 
-    // Throw BufferExceededException if all buffer frames are pinned
+    // set up frame and return the frame number
+    bufDescTable[clockHand].Set(bufDescTable[clockHand].file, bufDescTable[clockHand].pageNo);
+    frame  = bufDescTable[clockHand].frameNo;
```
Comment (rileyjsumner) [inline src/buffer.cpp:93]: Since frame is a pointer does this need to be `&bufDescTable` ? I'm not sure how cpp works lol

File: src/buffer.cpp [hunk 3]
```diff
@@ -101,19 +105,20 @@ void BufMgr::allocBuf(FrameId& frame) {
...
-    } catch (HashNotFoundException err) {
+        hashTable.lookup(file, pageNo, frameId); // here
+    } catch (HashNotFoundException const&) {
+
         // new frame is allocated from the buffer pool for reading page not present
         allocBuf(frameId);
         Page pageTemp = file.readPage(pageNo);
         bufPool[frameId] = pageTemp;
-        hashTable->insert(file, pageNo, frameId);
+        hashTable.insert(file, pageNo, frameId); // here
```
Comment (rileyjsumner) [inline src/buffer.cpp:121]: this here should be bufDescTable instead of hashTable as well?

File: src/buffer.cpp [hunk 4]
```diff
@@ -138,14 +143,14 @@ void BufMgr::readPage(File& file, const PageId pageNo, Page*& page) {
...
+    FrameId frameId = bufDescTable[clockHand].frameNo; // fetch the current frame Id
     bool isThrown = false;// default value is false, made true if exception is thrown
-    
+
     // wrapping the page-fetching process inside a try-catch block to deal with the case when the page is not found
     try {
         // fetch pointer to page we're trying to unpin
-        hashTable->lookup(file, pageNo, frameId);
-    } catch(HashNotFoundException e) {
+        hashTable.lookup(file, pageNo, frameId); // here
```
Comment (rileyjsumner) [inline src/buffer.cpp:152]: same here, bufDescTable right?

File: src/buffer.cpp [hunk 5]
```diff
@@ -174,23 +179,29 @@ void BufMgr::unPinPage(File& file, const PageId pageNo, const bool dirty) {
...
 
-    this->hashTable.Set(file, pageNo)
+    // create frameId store frame number returned from allocation in buffer
+    FrameId frameNo;
+    allocBuf(frameNo) ;// I think this is how I'm supposed to do this...
 
-    // set pageNo and page?
-    pageNo = &bufPool[frameNo].pageNo;
+    // add the page in the proper buffer frame and set the page pointer to this new page in the buffer
+    bufPool[frameNo] = newPage;
```
Comment (rileyjsumner) [inline src/buffer.cpp:192]: &newPage if pointer?

File: src/buffer.cpp [hunk 6]
```diff
@@ -174,23 +179,29 @@ void BufMgr::unPinPage(File& file, const PageId pageNo, const bool dirty) {
...
-    } catch(InsufficentSpaceException e) {
-
-    } catch(Exception e) {
-        e.printStackTrace();
-    }
+    // allocate the new page and then get the page number from it
+    Page newPage = file.allocatePage(); // here
+    pageNo = newPage.page_number(); // here
 
-    this->hashTable.Set(file, pageNo)
```
Comment (rileyjsumner) [inline src/buffer.cpp]: bufDescTable?

File: src/buffer.cpp [hunk 7]
```diff
@@ -174,23 +179,29 @@ void BufMgr::unPinPage(File& file, const PageId pageNo, const bool dirty) {
...
+    bufPool[frameNo] = newPage;
     page = &bufPool[frameNo];
+
+    // call set function to set up new frame in buffer
+    bufDescTable[frameNo].Set(file, pageNo);
+
+    //finally insert into hashtable, catching any exceptions
+
+    try {
+        hashTable.insert(file, pageNo, frameNo); // here
```
Comment (rileyjsumner) [inline src/buffer.cpp:201]: bufDescTable?

File: src/buffer.cpp [hunk 8]
```diff
@@ -210,17 +221,19 @@ void BufMgr::flushFile(File& file) {
...
-                bufDescTable[index].file -> writePage(bufPool[index]);
+                bufDescTable[index].file.writePage(bufPool[index]); // here
                 bufDescTable[index].dirty = false;
                 // then remove page from hashtable and bufDescTable
-                hashTable -> remove(file, bufDescTable[index].pageNo);
-                bufDescTable[index].Clear();
             }
+
+            //finally remove page from hashtable and the bufdesctable
+            hashTable.remove(file, bufDescTable[index].pageNo); // here
```
Comment (rileyjsumner) [inline src/buffer.cpp:235]: bufDescTable? although it is a param so maybe not

File: src/buffer.cpp [hunk 9]
```diff
@@ -235,15 +248,15 @@ void BufMgr::flushFile(File& file) {
 void BufMgr::disposePage(File& file, const PageId PageNo) {
     FrameId frameId; // stores frame ID from lookup call
     try{
-        hashTable -> lookup(file, PageNo, frameId);
+        hashTable.lookup(file, PageNo, frameId); // here
```
Comment (rileyjsumner) [inline src/buffer.cpp:251]: bufDescTable?

File: src/buffer.cpp (hunk 2 above)
Comment (dfkelly2) [reply src/buffer.cpp:93]: Yup I think you're right, think it should be &bufdesc...

**Pull request merged**